ext/hash/hash.c

```diff
@@ -111,6 +123,229 @@ PHP_HASH_API int php_hash_copy(const void *ops, void *orig_context, void *dest_c
+
+   The serialization result is an array. Each integer is serialized as a
+   32-bit integer, except that a run of 2 or more bytes is encoded as a
+   string, and each 64-bit integer is serialized as two 32-bit integers, least
+   significant bits first. This allows 32-bit and 64-bit architectures to
+   interchange serialized HashContexts. */
+
+PHP_HASH_API int php_hash_serialize_spec(const php_hashcontext_object *hash, zend_long *magic, zval *zv, const char *spec) /* {{{ */
+{
+	size_t pos = 0, max_alignment = 1, sz, count;
+	unsigned char *buf = (unsigned char *) hash->context;
+	zval tmp;
+	*magic = PHP_HASH_SERIALIZE_MAGIC_SPEC;
+	array_init(zv);
+	while (*spec != '\0' && *spec != '.') {
+		char specch = *spec;
+		count = parse_serialize_spec(&spec, &pos, &sz, &max_alignment);
+		if (pos + count * sz > hash->ops->context_size) {
+			return FAILURE;
+		}
+		if (specch == '-') {
+			pos += count;
+		} else if (sz == 1 && count > 1) {
+			ZVAL_STRINGL(&tmp, (char *) buf + pos, count);
+			zend_hash_next_index_insert(Z_ARRVAL_P(zv), &tmp);
+			pos += count;
+		} else {
+			while (count > 0) {
+				uint64_t val = one_from_buffer(sz, buf + pos);
+				pos += sz;
+				ZVAL_LONG(&tmp, (int32_t) val);
+				zend_hash_next_index_insert(Z_ARRVAL_P(zv), &tmp);
+				if (sz == 8) {
+					ZVAL_LONG(&tmp, (int32_t) (val >> 32));
+					zend_hash_next_index_insert(Z_ARRVAL_P(zv), &tmp);
+				}
+				--count;
+			}
+		}
+	}
+	if (*spec == '.' && align_to(pos, max_alignment) != hash->ops->context_size) {
+		return FAILURE;
+	}
+	return SUCCESS;
+}
+/* }}} */
+
+/* Unserialize a hash context serialized by `php_hash_serialize_spec` with `spec`.
+   Returns SUCCESS on success and a negative error code on failure.
+   Codes: FAILURE (-1) == generic failure
+   -999 == spec wrong size for context
+   -1000 - POS == problem at byte offset POS */
+
+PHP_HASH_API int php_hash_unserialize_spec(php_hashcontext_object *hash, zend_long magic, const zval *zv, const char *spec) /* {{{ */
+{
+	size_t pos = 0, max_alignment = 1, sz, count, j = 0;
+	unsigned char *buf = (unsigned char *) hash->context;
+	zval *elt;
+	if (magic != PHP_HASH_SERIALIZE_MAGIC_SPEC || Z_TYPE_P(zv) != IS_ARRAY) {
+		return FAILURE;
+	}
+	while (*spec != '\0' && *spec != '.') {
+		char specch = *spec;
+		count = parse_serialize_spec(&spec, &pos, &sz, &max_alignment);
+		if (pos + count * sz > hash->ops->context_size) {
+			return -999;
+		}
+		if (specch == '-') {
+			pos += count;
+		} else if (sz == 1 && count > 1) {
+			elt = zend_hash_index_find(Z_ARRVAL_P(zv), j);
+			if (!elt || Z_TYPE_P(elt) != IS_STRING || Z_STRLEN_P(elt) != count) {
+				return -1000 - pos;
+			}
+			++j;
+			memcpy(buf + pos, Z_STRVAL_P(elt), count);
+			pos += count;
+		} else {
+			while (count > 0) {
+				uint64_t val;
+				elt = zend_hash_index_find(Z_ARRVAL_P(zv), j);
+				if (!elt || Z_TYPE_P(elt) != IS_LONG) {
+					return -1000 - pos;
+				}
+				++j;
+				val = (uint32_t) zval_get_long(elt);
```

Comment:
```suggestion
				val = (uint32_t) Z_LVAL_P(elt);
```

zval_get_long is a casting function.